### PR TITLE
`VertexOrchestrator` apply node selector constraint if `gpu_limit > 0`

### PR DESCRIPTION
**Pull request by gabrielmbmb**
## Describe changes
When using the `VertexOrchestrator` and providing a value for the `node_selector_constraint` to get a GPU, all the steps will get a GPU even if it is not needed for that step. This is undesired, because you may want the training step to use GPU, but not all steps in the pipeline.

With this PR you can do something like:

```python
from zenml.steps import step, ResourceConfiguration

@step(resource_configuration=ResourceConfiguration(cpu_count=4.0, gpu_count=0, memory="64GB"))
def preprocess() -> int:
    ...

@step(resource_configuration=ResourceConfiguration(cpu_count=4.0, gpu_count=1, memory="64GB"))
def train(a: int, b:int) -> int:
    ...
```
and expect only the `train` step to request the GPU specified in the `node_selector_constraint` attribute to Vertex.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/mlops-stacks/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 1]
```diff
@@ -115,11 +118,14 @@ def _validate_stack_requirements(stack: "Stack") -> Tuple[bool, str]:
...
             if container_registry and container_registry.config.is_local:
-                return False, (
-                    f"The Vertex orchestrator does not support local "
-                    f"container registries. You should replace the component '"
-                    f"{container_registry.name}' "
-                    f"{container_registry.type.value} to a remote one."
+                return (
+                    False,
+                    (
+                        "The Vertex orchestrator does not support local"
```
Comment (htahir1) [inline src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:124]: Any reason for this formatting change? Im trying to see if something changed but dont notice it

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 2]
```diff
@@ -131,15 +137,18 @@ def _validate_stack_requirements(stack: "Stack") -> Tuple[bool, str]:
...
-                    f"local stack component. The Vertex AI Pipelines "
-                    f"orchestrator requires that all the components in the "
-                    f"stack used to execute the pipeline have to be not local, "
-                    f"because there is no way for Vertex to connect to your "
-                    f"local machine. You should use a flavor of "
-                    f"{stack_comp.type.value} other than '"
-                    f"{stack_comp.flavor}'."
+                return (
+                    False,
+                    (
```
Comment (htahir1) [inline src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:142]: Same as above

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 3]
```diff
@@ -149,15 +158,19 @@ def _validate_stack_requirements(stack: "Stack") -> Tuple[bool, str]:
...
-                return False, (
-                    f"The attribute `pipeline_root` has not been set and it "
-                    f"cannot be generated using the path of the artifact store "
-                    f"because it is not a "
-                    f"`zenml.integrations.gcp.artifact_store.GCPArtifactStore`."
-                    f" To solve this issue, set the `pipeline_root` attribute "
-                    f"manually executing the following command: "
-                    f"`zenml orchestrator update {stack.orchestrator.name} "
-                    f'--pipeline_root="<Cloud Storage URI>"`.'
+                return (
```
Comment (htahir1) [inline src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:161]: Same as above

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 4]
```diff
@@ -115,11 +118,14 @@ def _validate_stack_requirements(stack: "Stack") -> Tuple[bool, str]:
             # Validate that the container registry is not local.
             container_registry = stack.container_registry
             if container_registry and container_registry.config.is_local:
-                return False, (
-                    f"The Vertex orchestrator does not support local "
-                    f"container registries. You should replace the component '"
-                    f"{container_registry.name}' "
-                    f"{container_registry.type.value} to a remote one."
+                return (
```
Comment (fa9r) [inline src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:121]: This seems to be identical to the old version. If so, I'd prefer to keep the old formatting, which looked cleaner IMO.

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 5]
```diff
@@ -131,15 +137,18 @@ def _validate_stack_requirements(stack: "Stack") -> Tuple[bool, str]:
...
-                    f"The '{stack_comp.name}' {stack_comp.type.value} is a "
-                    f"local stack component. The Vertex AI Pipelines "
-                    f"orchestrator requires that all the components in the "
-                    f"stack used to execute the pipeline have to be not local, "
-                    f"because there is no way for Vertex to connect to your "
-                    f"local machine. You should use a flavor of "
-                    f"{stack_comp.type.value} other than '"
-                    f"{stack_comp.flavor}'."
+                return (
+                    False,
```
Comment (fa9r) [inline src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:141]: Same as above, old formatting looked cleaner IMO.

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 6]
```diff
@@ -149,15 +158,19 @@ def _validate_stack_requirements(stack: "Stack") -> Tuple[bool, str]:
...
-                    f"The attribute `pipeline_root` has not been set and it "
-                    f"cannot be generated using the path of the artifact store "
-                    f"because it is not a "
-                    f"`zenml.integrations.gcp.artifact_store.GCPArtifactStore`."
-                    f" To solve this issue, set the `pipeline_root` attribute "
-                    f"manually executing the following command: "
-                    f"`zenml orchestrator update {stack.orchestrator.name} "
-                    f'--pipeline_root="<Cloud Storage URI>"`.'
+                return (
+                    False,
```
Comment (fa9r) [inline src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:162]: Same as above, old formatting looked cleaner IMO.

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 7]
```diff
@@ -232,16 +245,32 @@ def _configure_container_resources(
...
         gpu_limit = resource_settings.gpu_count or self.config.gpu_limit
         if gpu_limit is not None:
             container_op = container_op.set_gpu_limit(gpu_limit)
 
+        if self.config.node_selector_constraint is not None:
+            constraint_label = self.config.node_selector_constraint[0]
+            value = self.config.node_selector_constraint[1]
+            if (
+                (
+                    constraint_label
```
Comment (htahir1) [inline src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:257]: In this case the person has set `gpu_count=None` explicitly? Should we allow this case?

**Review by htahir1 — COMMENTED** [review]
Left a few comments but looks like a good catch and very valuable addition to the Vertex orchestator!

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py (hunk 1 above)
Comment (gabrielmbmb) [reply src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:124]: Hey, I don't know! Let me check if I messed up something with `black`

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 8]
```diff
@@ -232,16 +245,32 @@ def _configure_container_resources(
...
+                )
+                or (
+                    constraint_label
+                    == GKE_ACCELERATOR_NODE_SELECTOR_CONSTRAINT_LABEL
+                    and gpu_limit is not None
+                    and gpu_limit > 0
+                )
+                or constraint_label
+                != GKE_ACCELERATOR_NODE_SELECTOR_CONSTRAINT_LABEL
+            ):
```
Comment (fa9r) [inline src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:269]: Could be done a bit shorter:

```suggestion
            if not (
                constraint_label
                == GKE_ACCELERATOR_NODE_SELECTOR_CONSTRAINT_LABEL
                and gpu_limit == 0
            ):
```

**Review by fa9r — CHANGES_REQUESTED** [review]
Hi Gabriel, thanks for the PR!

`gpu_count` being non-negative instead of positive certainly makes sense, this would have been an issue with other orchestrators and step operators as well. Great catch!

Overall the PR looks good to me, I just left a few suggestions for formatting changes :)

**Comment by gabrielmbmb** [comment]
Sorry about the formatting guys! I had the `--experimental-string-processing` argument of `black` defined in my VSCode configuration

File: src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py [hunk 9]
```diff
@@ -232,16 +235,22 @@ def _configure_container_resources(
...
 
+        if self.config.node_selector_constraint is not None:
+            constraint_label = self.config.node_selector_constraint[0]
+            value = self.config.node_selector_constraint[1]
+            if not (
+                constraint_label
+                == GKE_ACCELERATOR_NODE_SELECTOR_CONSTRAINT_LABEL
+                and gpu_limit == 0
+            ):
+                container_op.add_node_selector_constraint(
```
[the remark this answers is not in the document]
Comment (fa9r) [reply src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:250]: well I guess not necessary if the it's the last command in the method, nvm

**Review by htahir1 — APPROVED** [review]

**Pull request merged**